prompting/apparmor/message.go

```diff
@@ -0,0 +1,436 @@
+package apparmor
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+
+	"github.com/snapcore/snapd/arch"
+)
+
+// Message fields are defined as raw sized integer types as the same type may be
+// packed as 16 bit or 32 bit integer, to accommodate other fields in the
+// structure.
+
+// MsgHeader is the header of all apparmor notification messages.
+//
+// The header encodes the size of the entire message, including variable-size
+// components and the version of the notification protocol.
+type MsgHeader struct {
+	// Length is the length of the entire message, including any more
+	// specialized messages appended after the header and any variable-length
+	// objects referenced by any such message.
+	Length uint16
+	// Version is the version of the communication protocol.
+	// Currently version 2 is implemented by the demo kernel based on Linux 5.7.7,
+	// but version 3 was used in the past.
+	Version uint16
+}
+
+// UnmarshalBinary unmarshals the header from binary form.
+func (msg *MsgHeader) UnmarshalBinary(data []byte) error {
+	const prefix = "cannot unmarshal apparmor message header"
+
+	// Unpack fixed-size elements.
+	order := arch.Endian() // ioctl messages are native byte order, but others might not be
+	buf := bytes.NewBuffer(data)
+	if err := binary.Read(buf, order, msg); err != nil {
+		return fmt.Errorf("%s: %s", prefix, err)
+	}
+
+	if msg.Version != 2 {
+		return fmt.Errorf("%s: unsupported version: %d", prefix, msg.Version)
+	}
+	if int(msg.Length) != len(data) {
+		return fmt.Errorf("%s: length mismatch %d != %d",
+			prefix, msg.Length, len(data))
+	}
+
+	return nil
+}
+
+// RequestBuffer returns a new buffer for communication with the kernel.
+// The buffer contains encoded information about its size and protocol version.
```

Comment:
I don't understand from this explanation how this is supposed to be used. Also again should this return a dedicated type?

Reply:
Good question. This function will be called from the notifier run loop. It creates/prepares a buffer, and the notifier then immediately passes it into `NotifyIoctl`, which performs an ioctl syscall and fills the buffer with a message from apparmor/kernel.

Then the notifier passes the buffer into`decodeAndDispatchRequest()`, which is responsible for parsing the message, extracting the access request, and sending it on to be handled by the prompt listener. It also forks a goroutine to wait for a response from the listener, and then send a message back to apparmor.

I think this (get buffer and send ioctl request) could be combined into one function, probably in `ioctl.go`, which allocates the buffer, configures it, and then makes the syscall. This is because `RequestBuffer` configures the buffer for an ioctl syscall anyway, and that is its only intended use case. I think I'll rename this function to `PrepareIoctlRequestBuffer`, and then create a new function `GetApparmorMessage()` which calls `PrepareIoctlRequestBuffer()` and then `NotifyIoctl()` using that buffer. Does this sound good to you?

Thanks!

Reply:
I added the `IoctlRequestBuffer` type, and replaced the `RequestBuffer()` with `NewIoctlRequestBuffer()`. I replaced the `[]byte` parameter to `NotifyIoctl()` with a `IoctlRequestBuffer` parameter. I also added the `ReceiveApparmorMessage()` function, which calls `NewIoctlRequestBuffer()` followed by `NotifyIoctl()`, and returns the final buffer with the apparmor message from the kernel.